File: examples/orbital.exs
defmodule Orbital do
  def go do
    state = initialize("Orbital")
    preflight state
    Process.sleep 100
    launch state
  end

  def initialize(name) do
    {:ok, conn} = Jooce.start_link(name)
    {:ok, vessel_id, _} = Jooce.SpaceCenter.active_vessel(conn)
    {:ok, autopilot_pid} = Jooce.Controller.Autopilot.start_link(conn, vessel_id)
    {:ok, control_pid} = Jooce.Controller.Control.start_link(conn, vessel_id)
    {:ok, flight_pid} = Jooce.Controller.Flight.start_link(conn, vessel_id)
    {:ok, resources_pid} = Jooce.Controller.Resources.start(conn, vessel_id)

    %{conn: conn, vessel_id: vessel_id, autopilot_pid: autopilot_pid, control_pid: control_pid, flight_pid: flight_pid, resources_pid: resources_pid}
  end

  def preflight(state) do
    Jooce.Controller.Autopilot.pitch(state.autopilot_pid, 90.0)
    Jooce.Controller.Autopilot.heading(state.autopilot_pid, 90.0)
    Jooce.Controller.Autopilot.on(state.autopilot_pid)
    Jooce.Controller.Control.throttle(state.control_pid, 1.0)
  end

  def launch(state) do
    IO.puts "Launch"
    Jooce.Controller.Control.stage(state.control_pid)
    ascent_phase(state)
  end

  def ascent_phase(state) do
    altitude = Jooce.Controller.Flight.surface_altitude(state.flight_pid)
    cond do
      altitude < 500 ->
        ascent_phase(state)
      true ->
        Jooce.Controller.Autopilot.pitch(state.autopilot_pid, 60.0)
        gravity_turn(state)
    end
  end

  def gravity_turn(state) do
    fuel = Jooce.Controller.Resources.liquid_fuel(state.resources_pid)
    cond do
      fuel <= 0.1 ->
        IO.puts "Launch stage separation"
        Jooce.Controller.Control.throttle(state.control_pid, 0.0)
        Process.sleep 100
        Jooce.Controller.Control.stage(state.control_pid)
        Jooce.Controller.Autopilot.off(state.autopilot_pid)
        Jooce.Controller.Autopilot.sas_on(state.autopilot_pid)
        Process.sleep 100
        Jooce.Controller.Autopilot.sas_mode(state.autopilot_pid, :prograde)
        coast_to_apoapsis(state)
      true ->
        # IO.puts "Fuel: #{fuel}"
        gravity_turn(state)
    end
  end

  # def adjust_gravity_turn(state) do
  #   send state.flight_pid, {:altitude, :mean, self()}
  #   receive do
  #     {:ok, altitude} ->
  #       angle = 90.0 - (:math.pow(altitude / 90_000, 0.4) * 90.0)
  # #       # angle = cond do
  # #       #           angle < 0.01 ->
  # #       #             0.01
  # #       #           angle > 89.9 ->
  # #       #             89.9
  # #       #           true ->
  # #       #             angle
  # #       #         end
  #       IO.puts "#{altitude}: #{angle}"
  # #       send state.autopilot_pid, {:pitch, angle}
  # #       Process.sleep 1000
  #       gravity_turn(state)
  #   end
  # end

  def coast_to_apoapsis(state, altitude \\ 0) do
    new_altitude = Jooce.Controller.Flight.mean_altitude(state.flight_pid)
    cond do
      new_altitude <= 80_000 ->
        coast_to_apoapsis(state, new_altitude)
      new_altitude < altitude ->
        Jooce.Controller.Autopilot.sas_on(state.autopilot_pid)
        Process.sleep 100
        Jooce.Controller.Autopilot.sas_mode(state.autopilot_pid, :retrograde)
        descent_phase(state)
      true ->
        Jooce.Controller.Autopilot.sas_off(state.autopilot_pid)
        coast_to_apoapsis(state, new_altitude)
    end
  end

  def descent_phase(state) do
    altitude = Jooce.Controller.Flight.mean_altitude(state.flight_pid)
    cond do
      altitude > 60_000 ->
        descent_phase(state)
      altitude > 5_000 ->
        Jooce.Controller.Autopilot.sas_off(state.autopilot_pid)
        descent_phase(state)
      true ->
        IO.puts "Deploying parachute"
        Jooce.Controller.Control.stage(state.control_pid)
        landing_phase(state)
    end
  end

  def landing_phase(state) do
    altitude = Jooce.Controller.Flight.surface_altitude(state.flight_pid)
    cond do
      altitude > 0.1 ->
        landing_phase(state)
      true ->
        IO.puts("Landed")
        # Jooce.stop(state.conn)
    end
  end
end

Orbital.go

File: examples/orbital.py
# -*- coding: utf-8 -*-
"""kRPC script to put a command pod into orbit.

Based on the sample code at https://krpc.github.io/krpc/tutorials/launch-into-orbit.html
with modifications for our own rocket, as well as a few tweaks.

"""
import math
import time
import krpc

##
## constants
##
turn_start_altitude = 0
turn_end_altitude = 90000
target_altitude = 90000
max_q = 20000
orbits = 3

##
## variables for max Q PID
##
k_p = 2.0
k_d = 1.0
throttle_pid_d_accum = 0

##
## connect to KSP
##
conn = krpc.connect(name = 'orbital.py')
vessel = conn.space_center.active_vessel

##
## set up streams for telemetry
##
ut = conn.add_stream(getattr, conn.space_center, 'ut')
altitude = conn.add_stream(getattr, vessel.flight(), 'mean_altitude')
q = conn.add_stream(getattr, vessel.flight(), 'dynamic_pressure')
apoapsis = conn.add_stream(getattr, vessel.orbit, 'apoapsis_altitude')
stage_1_resources = vessel.resources_in_decouple_stage(stage = 4, cumulative = False)
stage_1_fuel = conn.add_stream(stage_1_resources.amount, 'LiquidFuel')
stage_2_resources = vessel.resources_in_decouple_stage(stage = 2, cumulative = False)
stage_2_fuel = conn.add_stream(stage_2_resources.amount, 'LiquidFuel')
latitude = conn.add_stream(getattr, vessel.flight(), 'latitude')
longitude = conn.add_stream(getattr, vessel.flight(), 'longitude')

##
## pre-launch setup
##
vessel.control.sas = False
vessel.control.rcs = False
vessel.control.throttle = 1.0

ksc_longitude = longitude()
deorbit_longitude = ksc_longitude - 49.0

##
## launch
##
print('Launch')
vessel.control.activate_next_stage()
vessel.auto_pilot.engage()
vessel.auto_pilot.target_pitch_and_heading(90, 90)

##
## ascent phase
##
stage_1_separated = False
turn_angle = 0
while True:

    # Gravity turn
    if altitude() > turn_start_altitude and altitude() < turn_end_altitude:
        frac = ((altitude() - turn_start_altitude) / (turn_end_altitude - turn_start_altitude)) ** 0.4
        new_turn_angle = frac * 90
        if abs(new_turn_angle - turn_angle) > 0.5:
            turn_angle = new_turn_angle
            vessel.auto_pilot.target_pitch_and_heading(90 - turn_angle, 90)

    # Separate first stage when finished
    if not stage_1_separated:
        if stage_1_fuel() < 0.1:
            print('MECO')
            vessel.control.throttle = 0.0
            time.sleep(1)
            print('First stage separation')
            vessel.control.activate_next_stage()
            stage_1_separated = True
            time.sleep(1)
            vessel.control.activate_next_stage()
            vessel.control.throttle = 1.0

    # Decrease throttle when approaching target apoapsis
    if apoapsis() > target_altitude * 0.9:
        break

    # throttle back if Q is too large
    q_error = max_q - q()
    p_value = k_p * q_error
    d_value = k_d * (q_error - throttle_pid_d_accum)
    throttle_pid_d_accum = q_error
    new_throttle_value = 1.0 + ((p_value + d_value) / 40000)
    if new_throttle_value > 1.0:
        new_throttle_value = 1.0
    elif new_throttle_value < 0.0:
        new_throttle_value = 0.0
    vessel.control.throttle = new_throttle_value

##
## disable engines when target apoapsis is reached
##
vessel.control.throttle = 0.25
while apoapsis() < target_altitude:
    if not stage_1_separated:
        if stage_1_fuel() < 0.1:
            print('MECO')
            vessel.control.throttle = 0.0
            time.sleep(1)
            print('First stage separation')
            vessel.control.activate_next_stage()
            stage_1_separated = True
            time.sleep(1)
            vessel.control.activate_next_stage()
            vessel.control.throttle = 0.25
    pass
vessel.control.throttle = 0.0
time.sleep(1)

##
## dump the first stage if we still have it
##
if not stage_1_separated:
    print('MECO')
    vessel.control.throttle = 0.0
    time.sleep(1)
    print('First stage separation')
    vessel.control.activate_next_stage()
    stage_1_separated = True
    time.sleep(1)
    vessel.control.activate_next_stage()
    vessel.control.throttle = 0.0

##
## wait until out of atmosphere
##
print('Coasting out of atmosphere')
while altitude() < 70500:
    pass

print('Circularizing')

##
## plan circularization burn (using vis-viva equation)
##
mu = vessel.orbit.body.gravitational_parameter
r = vessel.orbit.apoapsis
a1 = vessel.orbit.semi_major_axis
a2 = r
v1 = math.sqrt(mu * ((2.0 / r) - (1.0 / a1)))
v2 = math.sqrt(mu * ((2.0 / r) - (1.0 / a2)))
delta_v = v2 - v1
node = vessel.control.add_node(ut() + vessel.orbit.time_to_apoapsis, prograde = delta_v)

##
## calculate burn time (using rocket equation)
##
F = vessel.available_thrust
Isp = vessel.specific_impulse * 9.82
m0 = vessel.mass
m1 = m0 / math.exp(delta_v / Isp)
flow_rate = F / Isp
burn_time = (m0 - m1) / flow_rate

##
## orient ship
##
vessel.auto_pilot.disengage()
vessel.control.sas = True
time.sleep(1)
vessel.control.sas_mode = conn.space_center.SASMode.maneuver
vessel.auto_pilot.wait()

##
## wait until burn
##
burn_ut = ut() + vessel.orbit.time_to_apoapsis - (burn_time / 2.0)
lead_time = 5
conn.space_center.warp_to(burn_ut - lead_time)

##
## execute burn
##
time_to_apoapsis = conn.add_stream(getattr, vessel.orbit, 'time_to_apoapsis')
while time_to_apoapsis() - (burn_time / 2.0) > 0:
    pass
vessel.control.throttle = 1.0
time.sleep(burn_time - 0.1)
vessel.control.throttle = 0.05
remaining_burn = conn.add_stream(node.remaining_burn_vector, node.reference_frame)
while remaining_burn()[1] > 0:
    pass
vessel.control.throttle = 0.0
node.remove()

##
## orbiting
##
print('In orbit')

vessel.auto_pilot.disengage()
vessel.control.sas = True
time.sleep(1)
vessel.control.sas_mode = conn.space_center.SASMode.prograde
vessel.auto_pilot.wait()

current_orbit = 1
print('Starting orbit %d' % current_orbit)
previous_longitude = ksc_longitude
current_longitude = longitude()

while True:
    if (previous_longitude < ksc_longitude) and (current_longitude >= ksc_longitude):
        current_orbit = current_orbit + 1
        print('Starting orbit %d' % current_orbit)
    if (current_orbit > orbits) and (previous_longitude < deorbit_longitude) and (current_longitude >= deorbit_longitude):
        break
    time.sleep(1)
    previous_longitude = current_longitude
    current_longitude = longitude()

##
## deorbiting
##
print('Deorbiting')

vessel.auto_pilot.disengage()
vessel.control.sas = True
time.sleep(1)
vessel.control.sas_mode = conn.space_center.SASMode.retrograde
vessel.auto_pilot.wait()
time.sleep(1)

vessel.control.throttle = 1.0
while stage_2_fuel() > 0.1:
    pass
print('SECO')
vessel.control.throttle = 0.0
time.sleep(1)
vessel.control.activate_next_stage()

while altitude() > 60000:
    time.sleep(1)

vessel.control.sas = False

while altitude() > 6000:
    pass

print('Deploying parachute')
vessel.control.activate_next_stage()

print('Final longitude: %.4f' % longitude())
